### PR TITLE
feat: add switch component

### PR DESCRIPTION
**Pull request by Sandtory**
- add the main Switch component, based off of Button component.
- employ React's forwardRef to obtain a reference to the underlying view.
- integrate custom hooks (useStyles, useToken) to derive theme-specific styles and animation timings.
- incorporated Paper component for the knob, adding depth with elevation.
- add sub component SmallSwitch with typeOf Switch.
- add SwitchScreen with examples of Switch usage.

 ^ written by HenningGPT

### CODE REVIEW

File: apps/chronicles/screens/components/SwitchScreen.tsx [hunk 1]
```diff
@@ -0,0 +1,101 @@
+import React, { useState } from "react";
+import { ScrollView, View } from "react-native";
+import { EDSStyleSheet, Spacer, Typography, useStyles } from "@equinor/mad-components";
+import { Switch } from "../../../../packages/components/src/components/Switch/";
```
Comment (HaakonSvane) [inline apps/chronicles/screens/components/SwitchScreen.tsx:4]: This import should be pointing to the package and not relative reference in the monorepo.

Comment (HaakonSvane) [reply apps/chronicles/screens/components/SwitchScreen.tsx:4]: It is probably why we dont get dark mode support btw!

File: apps/chronicles/screens/components/SwitchScreen.tsx [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+        </ScrollView>
+    );
+};
+
+const themeStyles = EDSStyleSheet.create(theme => ({
+    container: {
+        paddingHorizontal: theme.spacing.container.paddingHorizontal,
+        paddingVertical: theme.spacing.container.paddingVertical,
+    },
+    SwitchRow: {
```
Comment (HaakonSvane) [inline apps/chronicles/screens/components/SwitchScreen.tsx:89]: style names should not be capitalized

File: apps/chronicles/screens/components/SwitchScreen.tsx [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+        paddingVertical: 8,
```
Comment (HaakonSvane) [inline apps/chronicles/screens/components/SwitchScreen.tsx:90]: Instead of hard coding this value, if you want spacing between your components, try to use the `<Spacer />` component in mad components. It comes with some predefined values

File: apps/chronicles/screens/components/SwitchScreen.tsx [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+        flexDirection: "row",
+        justifyContent: "space-around",
+    },
+    switchDescription: {
+        flexDirection: "column",
+        alignItems: "center",
+        paddingHorizontal: 8,
```
Comment (HaakonSvane) [inline apps/chronicles/screens/components/SwitchScreen.tsx:97]: Same here as for my other comment on spacing

File: apps/chronicles/types.tsx [hunk 5]
```diff
@@ -17,6 +17,8 @@ export type DiscoverStackParamList = {
     Discover: undefined;
     Paper: undefined;
     Button: undefined;
+    Switch: undefined;
+    SmallSwitch: undefined;
```
Comment (HaakonSvane) [inline apps/chronicles/types.tsx:21]: We don't need a separate navigation route for the small switch since it is shown on the same screen as the big one

File: apps/chronicles/screens/components/SwitchScreen.tsx [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+                    <Switch
+                        active={isDisabledActiveSwitch}
+                        disabled={isDisabled}
+                        onChange={handleDisabledActiveState}
+                    />
+                    <Typography>{"You can't turn me off"}</Typography>
+                </View>
+            </View>
+            <Spacer />
+            <Typography variant="h2">Small</Typography>
```
Comment (HaakonSvane) [inline apps/chronicles/screens/components/SwitchScreen.tsx:67]: does the small Switch support disabled? If so, maybe show it here as an example as well

File: packages/components/src/components/Switch/Switch.tsx [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+        const inactiveBackgroundAnimation = Animated.timing(backgroundProgressValue, {
+            toValue: 0,
+            duration: token.timing.animation.slow,
+            useNativeDriver: true,
+            easing: Easing.inOut(Easing.ease),
+        });
+
+        const backgroundDisplacement = backgroundProgressValue.interpolate({
+            inputRange: [0, 1],
+            outputRange: [0.001, 1],
```
Comment (HaakonSvane) [inline packages/components/src/components/Switch/Switch.tsx:82]: Why was this value set to 0.001 ?

File: packages/components/src/components/Switch/Switch.tsx [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+            easing: Easing.inOut(Easing.ease),
+        });
+
+        const displacement = knobProgressValue.interpolate({
+            inputRange: [0, 1],
+            outputRange: [0, WIDTH - KNOB_SIZE * 1.7],
+            extrapolate: "clamp",
+        });
+
+        const activeBackgroundAnimation = Animated.timing(backgroundProgressValue, {
```
Comment (HaakonSvane) [inline packages/components/src/components/Switch/Switch.tsx:66]: Notice that this animation does exactly the same as the activeKnobAnimation does.
This means we dont need two animations (one for the knob and one for the background). Since the animated values that the knob needs are different than what the background needs, this is handled in the interpolate methods you have defined.

By removing activeBackgroundAnimation and inactiveBackgroundAnimation, you can reduce the code in this file by a good chunk 👍

File: packages/components/src/components/Switch/Switch.tsx [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+    return {
+        toggleContainer: {
+            flexDirection: "row",
+            alignItems: "center",
+        },
+        pressableContainer: {
+            height: HEIGHT,
+            width: WIDTH,
+            padding: 7,
+            borderRadius: 100,
```
Comment (HaakonSvane) [inline packages/components/src/components/Switch/Switch.tsx:169]: Just to safe, maybe set the border radius to be WIDTH / 2 so that it scales with size

File: .changeset/thick-papayas-promise.md [hunk 10]
```diff
@@ -0,0 +1,5 @@
+---
+"@equinor/mad-components": patch
+---
+
+Added new component Switch, child component SmallSwitch and a SwitchScreen
```
Comment (HaakonSvane) [inline .changeset/thick-papayas-promise.md:5]: I would rephrase this to
> Added `Switch` and `Switch.Small` components

the code blocks helps users understand that these are actual jsx tags that can be used.
Also, the changeset is written for the mad components package, so we don't need to mention any SwitchScreens since that belongs to the chronicles app.

**Review by HaakonSvane — CHANGES_REQUESTED** [review]
Super awesome. Great work. You should be really proud 🔥
I added some minor comments for things we should look into / correct before we merge. other than that, quality code dude 😎

File: apps/chronicles/screens/components/SwitchScreen.tsx (hunk 3 above)
Comment (Sandtory) [reply apps/chronicles/screens/components/SwitchScreen.tsx:90]: this was taken from the button component, should i change it there aswell?

File: packages/components/src/components/Switch/Switch.tsx (hunk 7 above)
Comment (Sandtory) [reply packages/components/src/components/Switch/Switch.tsx:82]: it was based on some old version where the value had to be greater than zero, it will be changed

**Review by HaakonSvane — APPROVED** [review]
Awesome job 👍
